File: code_samples/go/design_patterns_singleton.go
package main

import (
    "fmt"
)

// Design Patterns: Singleton
// AI/ML Training Sample

type Singleton struct {
    Data string
}

func NewSingleton() *Singleton {
    return &Singleton{
        Data: "",
    }
}

func (s *Singleton) Process(input string) {
    s.Data = input
}

func (s *Singleton) Validate() bool {
    return len(s.Data) > 0
}

func (s *Singleton) GetData() string {
    return s.Data
}

func main() {
    instance := NewSingleton()
    instance.Process("example")
    fmt.Printf("Data: %s\n", instance.GetData())
    fmt.Printf("Valid: %v\n", instance.Validate())
}

File: code_samples/scala/design_patterns_singleton.scala
/**
 * Design Patterns: Singleton
 * AI/ML Training Sample
 */

class Singleton {
  private var data: String = ""
  
  def process(input: String): Unit = {
    data = input
  }
  
  def getData: String = data
  
  def validate: Boolean = data.nonEmpty
}

object SingletonExample {
  def main(args: Array[String]): Unit = {
    val instance = new Singleton()
    instance.process("example")
    println(s"Data: ${instance.getData}")
    println(s"Valid: ${instance.validate}")
  }
}
